src/urllib3/response.py

```diff
@@ -770,8 +798,14 @@ def _fp_read(self, amt: int | None = None) -> bytes:
         c_int_max = 2**31 - 1
         if (
             (amt and amt > c_int_max)
-            or (self.length_remaining and self.length_remaining > c_int_max)
+            or (
+                amt is None
```

Comment:
@sethmlarson `read1` is a standard method provided by `io.BufferedReader`s which `http.client.HTTPResponse` also provides

this PR is mostly about getting `read1` supported, there are various cleanups that make sense once it's in there.  e.g. with compressed streams it doesn't make sense to me that `amt` is passed to `read` on the stream.  the decompressed data might be much larger and hence using `read1` would mean that the caller might get a response much earlier

> I'm not sure why `amt is None` was added here? The new branch `if one_read` doesn't use `amt`.

without it this "slow path" would get used for large responses even if `amt` is small

Reply:
@sethmlarson in case urllib3 or http.client is going to `read1` 2 GiB or more and get the overflow error, the code under this if block prevents this by passing the maximum safe amount to `read1`

The condition would be more correct with the new `amt is None` check even without the `read1` addition. Not having it had no adverse effect previously, but now it's either `amt is None` has to be added here or code on lines 804–805 has to become more complex.